agent/consul/discovery_chain_endpoint_test.go

```diff
@@ -242,3 +244,127 @@ func TestDiscoveryChainEndpoint_Get(t *testing.T) {
 		require.Equal(t, expect, resp)
 	}
 }
+
+func TestDiscoveryChainEndpoint_Get_BlockOnNoChange(t *testing.T) {
+	if testing.Short() {
+		t.Skip("too slow for testing.Short")
+	}
+
+	_, s1 := testServerWithConfig(t, func(c *Config) {
+		c.PrimaryDatacenter = "dc1"
+	})
+
+	codec := rpcClient(t, s1)
+	readerCodec := rpcClient(t, s1)
+	writerCodec := rpcClient(t, s1)
+
+	waitForLeaderEstablishment(t, s1)
+	testrpc.WaitForTestAgent(t, s1.RPC, "dc1")
+
+	{ // create one unrelated entry
+		var out bool
+		require.NoError(t, msgpackrpc.CallWithCodec(codec, "ConfigEntry.Apply", &structs.ConfigEntryRequest{
+			Datacenter: "dc1",
+			Entry: &structs.ServiceResolverConfigEntry{
+				Kind:           structs.ServiceResolver,
+				Name:           "unrelated",
+				ConnectTimeout: 33 * time.Second,
+			},
+		}, &out))
+		require.True(t, out)
+	}
+
+	run := func(t *testing.T, dataPrefix string) {
```

Comment:
So to properly use a blocking query you will always do at least 2 calls:
1) first request where you don't know the index, so it's effectively `index=` or `index=0`; this does not block and immediately returns `index=ZZZ`
2) subsequent request sends `index=ZZZ` and may block if nothing has changed since (1), and then ultimately returns when something changes or 1s elapses

The spamming goroutine is running in between (1) and (2), so by the time (2) runs the first time the index value may have already gone up, so (2) immediately returns because something changed.

This is done in this specific way due to an implementation wrinkle in how `errNotFound` code works. It can ONLY suppress a wakeup if the blocking query looping code during (2) is allowed to loop around one time first, which only happens if (2) actually blocks because the value we got during (1) (index=ZZZ) is still the index value at the very start of (2). So if the spam is timed poorly then the blocking query code here may have to wake up a few extra times due to losing the lottery and so we have to pad it out a bit.